src/trel_dnssd/trel_dnssd.cpp

```diff
@@ -0,0 +1,542 @@
+ */
+
+#if OTBR_ENABLE_TREL
+
+#define OTBR_LOG_TAG "TrelDns"
+
+#include "trel_dnssd/trel_dnssd.hpp"
+
+#include <net/if.h>
+
+#include <openthread/instance.h>
+#include <openthread/link.h>
+#include <openthread/platform/trel.h>
+
+#include "common/code_utils.hpp"
+#include "utils/string_utils.hpp"
+
+static const char kTrelServiceName[] = "_trel._udp";
+
+void trelDnssdInitialize(const char *aTrelNetif)
+{
+    otbr::TrelDnssd::TrelDnssd::GetInstance().Initialize(aTrelNetif);
+}
+
+void trelDnssdStartBrowse(void)
+{
+    otbr::TrelDnssd::TrelDnssd::GetInstance().StartBrowse();
+}
+
+void trelDnssdStopBrowse(void)
+{
+    otbr::TrelDnssd::TrelDnssd::GetInstance().StopBrowse();
+}
+
+void trelDnssdRegisterService(uint16_t aPort, const uint8_t *aTxtData, uint8_t aTxtLength)
+{
+    otbr::TrelDnssd::TrelDnssd::GetInstance().RegisterService(aPort, aTxtData, aTxtLength);
+}
+
+void trelDnssdRemoveService(void)
+{
+    otbr::TrelDnssd::TrelDnssd::GetInstance().RemoveService();
+}
+
+namespace otbr {
+
+namespace TrelDnssd {
+
+TrelDnssd *TrelDnssd::sTrelDnssd = nullptr;
+
+void TrelDnssd::Initialize(std::string aTrelNetif)
+{
+    mTrelNetif = aTrelNetif;
+
+    if (IsInitialized())
+    {
+        otbrLogDebug("Initialized on netif \"%s\"", aTrelNetif.c_str());
+        CheckTrelNetifReady();
+    }
+    else
+    {
+        otbrLogDebug("Not initialized");
+    }
+}
+
+void TrelDnssd::StartBrowse(void)
+{
+    VerifyOrExit(IsInitialized());
+
+    otbrLogDebug("Start browsing %s service ...", kTrelServiceName);
+
+    assert(mSubscriberId == 0);
+    mSubscriberId = mPublisher.AddSubscriptionCallbacks(
+        [this](const std::string &aType, const Mdns::Publisher::DiscoveredInstanceInfo &aInstanceInfo) {
+            OnTrelServiceDiscovered(aType, aInstanceInfo);
+        },
+
+        nullptr);
+
+    if (IsReady())
+    {
+        mPublisher.SubscribeService(kTrelServiceName, "");
+    }
+
+exit:
+    return;
+}
+
+void TrelDnssd::StopBrowse(void)
+{
+    VerifyOrExit(IsInitialized());
+
+    otbrLogDebug("Stop browsing %s service.", kTrelServiceName);
+    assert(mSubscriberId > 0);
+
+    mPublisher.RemoveSubscriptionCallbacks(mSubscriberId);
+    mSubscriberId = 0;
+
+    if (IsReady())
+    {
+        mPublisher.UnsubscribeService(kTrelServiceName, "");
+    }
+
+exit:
+    return;
+}
+
+void TrelDnssd::RegisterService(uint16_t aPort, const uint8_t *aTxtData, uint8_t aTxtLength)
+{
+    assert(aPort > 0);
+    assert(aTxtData != nullptr);
+
+    VerifyOrExit(IsInitialized());
+
+    otbrLogDebug("Register %s service: port=%u, TXT=%d bytes", kTrelServiceName, aPort, aTxtLength);
+    otbrDump(OTBR_LOG_DEBUG, OTBR_LOG_TAG, "TXT", aTxtData, aTxtLength);
+
+    if (mRegisterInfo.IsValid() && IsReady())
+    {
+        DoRemoveTrelService();
+    }
+
+    mRegisterInfo.Assign(aPort, aTxtData, aTxtLength);
+
+    if (IsReady())
+    {
+        DoRegisterTrelService();
+    }
+
+exit:
+    return;
+}
+
+void TrelDnssd::RemoveService(void)
```

Comment:
Good suggestion. Will do.

Reply:
On the other hand, I use `RegisterService` and `RemoveService` because we are also using `trelDnssdRegisterService` and `trelDnssdRemoveService`. May also need to rename those?